### PR TITLE
added a gsk::ColorStopBuilder

### PR DESCRIPTION
**Pull request by kavanmevada**
#499

### CODE REVIEW

File: gsk4/src/color_stop.rs [hunk 1]
```diff
@@ -2,8 +2,25 @@
 
 use gdk::RGBA;
 use glib::translate::*;
+use std::borrow::Cow;
 use std::fmt;
 
+#[derive(Debug, Default)]
+#[must_use = "The builder must be built to be used"]
```
Comment (bilelmoussaoui) [inline gsk4/src/color_stop.rs:9]: Please copy the auto generated documentation similar to the other builder types.

File: gsk4/src/color_stop.rs [hunk 2]
```diff
@@ -2,8 +2,25 @@
...
+pub struct ColorStopBuilder(Vec<ColorStop>);
+
+impl ColorStopBuilder {
+    pub fn stop(mut self, offset: f32, color: RGBA) -> Self {
+        self.0.push(ColorStop::new(offset, color));
+        self
+    }
+
+    #[must_use = "The Vec<ColorStop> returned by this builder should probably be used"]
+    pub fn build(self) -> Vec<ColorStop> {
```
Comment (bilelmoussaoui) [inline gsk4/src/color_stop.rs:19]: Same here

File: gsk4/src/color_stop.rs [hunk 3]
```diff
@@ -2,8 +2,25 @@
 
 use gdk::RGBA;
 use glib::translate::*;
+use std::borrow::Cow;
 use std::fmt;
 
+#[derive(Debug, Default)]
```
Comment (bilelmoussaoui) [inline gsk4/src/color_stop.rs:8]: You should also add a `builder` method to `ColorStop` so you can use it with `ColorStop::builder`

File: gsk4/src/color_stop.rs [hunk 4]
```diff
@@ -2,8 +2,25 @@
 
 use gdk::RGBA;
 use glib::translate::*;
+use std::borrow::Cow;
```
Comment (bilelmoussaoui) [inline gsk4/src/color_stop.rs:5]:
```suggestion
```

File: examples/custom_paintable/custom_paintable/imp.rs [hunk 5]
```diff
@@ -35,18 +35,18 @@ impl PaintableImpl for CustomPaintable {
...
             &graphene::Point::new(width as f32, height as f32),
-            &[
-                gsk::ColorStop::new(0.0, gdk::RGBA::RED),
-                gsk::ColorStop::new(0.15, gdk::RGBA::new(1.0, 127_f32 / 255_f32, 0.0, 1.0)),
-                gsk::ColorStop::new(0.3, gdk::RGBA::new(1.0, 1.0, 0.0, 1.0)),
-                gsk::ColorStop::new(0.45, gdk::RGBA::GREEN),
-                gsk::ColorStop::new(0.6, gdk::RGBA::BLUE),
-                gsk::ColorStop::new(
+            &gsk::ColorStop::builder()
+                .stop(0.0, gdk::RGBA::RED)
```
Comment (bilelmoussaoui) [inline examples/custom_paintable/custom_paintable/imp.rs:39]: Maybe this should be called `at` instead?

**Review by bilelmoussaoui — CHANGES_REQUESTED** [review]

File: gsk4/src/color_stop.rs (hunk 3 above)
Comment (kavanmevada) [reply gsk4/src/color_stop.rs]: wdym? `pub fn builder() -> ColorStopBuilder` also implemented in `ColorStop`.

File: gsk4/src/color_stop.rs [hunk 6]
```diff
@@ -20,6 +48,14 @@ impl ColorStop {
         }
     }
 
+    // rustdoc-stripper-ignore-next
+    /// Creates a new builder-pattern struct instance to construct [`ColorStop`] objects.
+    ///
+    /// This method returns an instance of [`ColorStopBuilder`](crate::builders::ColorStopBuilder) which can be used to create [`ColorStop`] objects.
```
Comment (bilelmoussaoui) [inline gsk4/src/color_stop.rs:54]: ColorStopBuilder has to be exported as part of the builders module, see lib.rs

File: gsk4/src/color_stop.rs [hunk 7]
```diff
@@ -37,3 +73,30 @@ impl fmt::Debug for ColorStop {
...
+        .at(0.667, gdk::RGBA::from_str("#0000FF").unwrap())
+        .at(1.0, gdk::RGBA::from_str("#FF0000").unwrap())
+        .build();
+
+    let t2 = &[
+        ColorStop::new(0.0, gdk::RGBA::from_str("#FF0000").unwrap()),
+        ColorStop::new(0.333, gdk::RGBA::from_str("#FF0000").unwrap()),
+        ColorStop::new(0.667, gdk::RGBA::from_str("#0000FF").unwrap()),
+        ColorStop::new(1.0, gdk::RGBA::from_str("#FF0000").unwrap()),
+    ];
```
Comment (bilelmoussaoui) [inline gsk4/src/color_stop.rs:93]:
```suggestion
    let t1 = ColorStop::builder()
        .at(0.0, gdk::RGBA::RED)
        .at(0.333, gdk::RGBA::RED)
        .at(0.667, gdk::RGBA::BLUE)
        .at(1.0, gdk::RGBA::RED)
        .build();

    let t2 = &[
        ColorStop::new(0.0, gdk::RGBA::RED),
        ColorStop::new(0.333, gdk::RGBA::RED),
        ColorStop::new(0.667, gdk::RGBA::BLUE),
        ColorStop::new(1.0, gdk::RGBA::RED),
    ];
```

File: examples/custom_paintable/custom_paintable/imp.rs [hunk 8]
```diff
@@ -35,18 +35,18 @@ impl PaintableImpl for CustomPaintable {
             &graphene::Rect::new(0_f32, 0_f32, width as f32, height as f32),
```
Comment (bilelmoussaoui) [inline examples/custom_paintable/custom_paintable/imp.rs:35]: Commit message should be
```
gsk: Add a ColorStopBuilder

Fixes #499

**Comment by sdroege** [comment]
What's the status here?

**Comment by bilelmoussaoui** [comment]
Closing in favour of https://github.com/gtk-rs/gtk4-rs/pull/1139

**Pull request closed without merging**